### PR TITLE
Don't be overly aggressive on stream failures and closing

### PR DESCRIPTION
**Pull request by zwoop**
Randall found this in our production. I'm thinking 1/threshold is a good limit here, such that you can at least calculate a reasonable percentage based on a large enough sample.

### CODE REVIEW

**Comment by zwoop** [comment]
@masaori335  Now that we're looking at this code, why is this x2 on the threshold?

```
    if (this->connection_state.get_stream_error_rate() > std::min(1.0, Http2::stream_error_rate_threshold * 2.0)) {
```

File: proxy/http2/Http2ClientSession.cc [hunk 1]
```diff
@@ -610,8 +610,8 @@ Http2ClientSession::state_process_frame_read(int event, VIO *vio, bool inside_fr
       ip_port_text_buffer ipb;
       const char *client_ip = ats_ip_ntop(get_client_addr(), ipb, sizeof(ipb));
       Warning("HTTP/2 session error client_ip=%s session_id=%" PRId64
-              " closing a connection, because its stream error rate (%f) is too high",
-              client_ip, connection_id(), this->connection_state.get_stream_error_rate());
+              " closing a connection, because its stream error rate (%f) exceeded the threshold (%f)",
```
Comment (zwoop) [inline proxy/http2/Http2ClientSession.cc]: This was only done to make the Warning() consistent with the other three places we do this check.

File: proxy/http2/Http2ConnectionState.h [hunk 2]
```diff
@@ -250,7 +250,8 @@ class Http2ConnectionState : public Continuation
   get_stream_error_rate() const
   {
     int total = get_stream_requests();
-    if (total > 0) {
+
+    if (total >= (1 / Http2::stream_error_rate_threshold)) {
```
Comment (zwoop) [inline proxy/http2/Http2ConnectionState.h]: The point of this is to require a minimum number of samples before we can calculate a reasonably trustworthy rate of failures. The smaller the threshold, the more samples needed. Probably not statistically safe, but this at least avoids the issues where if (in the default configs) any of the first 10 streams has an error, it's enough to close the connection.

**Comment by masaori335** [comment]
> @masaori335 Now that we're looking at this code, why is this x2 on the threshold?

@maskit could you recall? It comes from 9c09dbc118f7ce22ce6a89ac88c7532f83ee23b0.

**Comment by masaori335** [comment]
This fixes #5195

**Comment by maskit** [comment]
> why is this x2 on the threshold?

It could be 1.5, 3 or whatever, but 5 is probably too big. If stream error rate exceeds the threshold, the connection will be closed *gracefully*. If stream error rate exceeds 2x of the threshold, the connection will be closed *immediately*. This is why the error message is slightly different, and the one for 2x says "too high".

I'm ok with this change, but the reason I didn't check a number of requests was that I thought we want to close stupid clients that causes stream error on the first request.

File: proxy/http2/Http2ClientSession.cc (hunk 1 above)
Comment (maskit) [reply proxy/http2/Http2ClientSession.cc]: Adding configured threshold value is fine, but removing "too high" would make it hard to see whether the close was graceful or immediate.

**Comment by zwoop** [comment]
Cherry-picked to 8.1.x
Cherry-picked to v9.0.x branch.

**Pull request merged**